adapters/gamma/gamma.go

```diff
@@ -0,0 +1,222 @@
+		return nil, errors
+	}
+	if gammaExt.WebID == "" {
+		err = &errortypes.BadInput{
+			Message: "WebID is empty",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	thisURI := a.URI
+	thisURI = thisURI + "?id=" + gammaExt.PartnerID
+	thisURI = thisURI + "&zid=" + gammaExt.ZoneID
+	thisURI = thisURI + "&wid=" + gammaExt.WebID
+	thisURI = thisURI + "&bidid=" + imp.ID
+	thisURI = thisURI + "&hb=pbmobile"
+	if request.Device != nil {
+		thisURI = thisURI + "&device_ip=" + request.Device.IP
+		thisURI = thisURI + "&device_model=" + request.Device.Model
+		thisURI = thisURI + "&device_os=" + request.Device.OS
+		if request.Device.UA != "" {
+			thisURI = thisURI + "&device_ua=" + url.QueryEscape(request.Device.UA)
+		}
+		thisURI = thisURI + "&device_ifa=" + request.Device.IFA
+	}
+	if request.App != nil {
+		thisURI = thisURI + "&app_id=" + request.App.ID
+		thisURI = thisURI + "&app_bundle=" + request.App.Bundle
+		thisURI = thisURI + "&app_name=" + request.App.Name
+	}
+	headers := http.Header{}
+	headers.Add("Accept", "*/*")
+	headers.Add("x-openrtb-version", "2.5")
+	if gammaExt.PartnerID == "bad-request-id" {
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+	}
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	addHeaderIfNonEmpty(headers, "Connection", "keep-alive")
+	addHeaderIfNonEmpty(headers, "cache-control", "no-cache")
+	addHeaderIfNonEmpty(headers, "Accept-Encoding", "gzip, deflate")
+
+	return &adapters.RequestData{
+		Method:  "GET",
+		Uri:     thisURI,
+		Headers: headers,
+	}, errors
+}
+func (a *GammaAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		err := &errortypes.BadInput{
+			Message: "No impressions in the bid request",
+		}
+		errs = append(errs, err)
+		return nil, errs
+	}
+	validImpExists := false
+	for i := 0; i < len(request.Imp); i++ {
+		if request.Imp[i].Banner != nil {
+			bannerCopy := *request.Imp[i].Banner
+			if bannerCopy.W == nil && bannerCopy.H == nil && len(bannerCopy.Format) > 0 {
+				firstFormat := bannerCopy.Format[0]
+				bannerCopy.W = &(firstFormat.W)
+				bannerCopy.H = &(firstFormat.H)
+			}
+			request.Imp[i].Banner = &bannerCopy
+			validImpExists = true
+		} else if request.Imp[i].Video != nil {
+			validImpExists = true
+		} else {
+			err := &errortypes.BadInput{
+				Message: fmt.Sprintf("Gamma only supports banner and video media types. Ignoring imp id=%s", request.Imp[i].ID),
+			}
+			errs = append(errs, err)
+			request.Imp = append(request.Imp[:i], request.Imp[i+1:]...)
```

Comment:
Since function `append(arr[:i], arr[i+1]...)` appends the rest of the elements of the array at a cost of O(n), we can assume that the time complexity is O(n^2)
```
115     // Assume: request.Imp = []Imp{"banner","other","other","other","other","other","video","banner"}
116     validImpExists := false
117     for i := 0; i < len(request.Imp); i++ {
118         if request.Imp[i].Banner != nil {
119             bannerCopy := *request.Imp[i].Banner
120             if bannerCopy.W == nil && bannerCopy.H == nil && len(bannerCopy.Format) > 0 {
121                 firstFormat := bannerCopy.Format[0]
122                 bannerCopy.W = &(firstFormat.W)
123                 bannerCopy.H = &(firstFormat.H)
124             }
125             request.Imp[i].Banner = &bannerCopy
126             validImpExists = true
127         } else if request.Imp[i].Video != nil {
128             validImpExists = true
129         } else {
130             err := &errortypes.BadInput{
131                 Message: fmt.Sprintf("Gamma only supports banner and video media types. Ignoring imp id=%s", request.Imp[i].ID),
132             }
133             errs = append(errs, err)
134             request.Imp = append(request.Imp[:i], request.Imp[i+1:]...) //we'll have a cost of O(n) every time we find an Imp of the type "other"
135             i--                                                         //yielding a cost of O(n*n) or n^2
136         }
137     }
```
Do you think this approach can be improved upon? What about an index array, a boolean array, or, even better, a bitmap?

Reply:
Maybe something like:
```
diff --git a/adapters/gamma/gamma.go b/adapters/gamma/gamma.go
index 0d5fade..f8389bd 100644
--- a/adapters/gamma/gamma.go
+++ b/adapters/gamma/gamma.go
@@ -113,7 +113,7 @@ func (a *GammaAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapte
                errs = append(errs, err)
                return nil, errs
        }
-       validImpExists := false
+       var validImpIndex = make([]int, 0, len(request.Imp)) //len(invalidImpIndex) = 0, cap(invalidImpIndex) = 0
        for i := 0; i < len(request.Imp); i++ {
                if request.Imp[i].Banner != nil {
                        bannerCopy := *request.Imp[i].Banner
@@ -123,33 +123,43 @@ func (a *GammaAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapte
                                bannerCopy.H = &(firstFormat.H)
                        }
                        request.Imp[i].Banner = &bannerCopy
-                       validImpExists = true
-               } else if request.Imp[i].Video != nil {
-                       validImpExists = true
-               } else {
+               } else if request.Imp[i].Video == nil {
                        err := &errortypes.BadInput{
                                Message: fmt.Sprintf("Gamma only supports banner and video media types. Ignoring imp id=%s", request.Imp[i].ID),
                        }
                        errs = append(errs, err)
-                       request.Imp = append(request.Imp[:i], request.Imp[i+1:]...)
-                       i--
+                       invalidImpIndex = append(invalidImpIndex, i)
                }
        }

-       if !validImpExists {
+       var adapterRequests []*adapters.RequestData
+       var j int = 0
+       if len(invalidImpIndex) == 0 { //Proceed as usual
+               for _, imp := range request.Imp {
+                       adapterReq, errors := a.makeRequest(request, imp)
+                       if adapterReq != nil {
+                               adapterRequests = append(adapterRequests, adapterReq)
+                       }
+                       errs = append(errs, errors...)
+               }
+       } else if len(request.Imp) == len(invalidImpIndex) { //only true if every Imp was not a Banner or a Video
                err := &errortypes.BadInput{
                        Message: fmt.Sprintf("No valid impression in the bid request"),
                }
                errs = append(errs, err)
                return nil, errs
-       }
-       var adapterRequests []*adapters.RequestData
-       for _, imp := range request.Imp {
-               adapterReq, errors := a.makeRequest(request, imp)
-               if adapterReq != nil {
-                       adapterRequests = append(adapterRequests, adapterReq)
+       } else {
+               for i := 0; i < len(request.Imp); i++ {
+                       if i == invalidImpIndex[j] {
+                               j++
+                       } else {
+                               adapterReq, errors := a.makeRequest(request, request.Imp[i])
+                               if adapterReq != nil {
+                                       adapterRequests = append(adapterRequests, adapterReq)
+                               }
+                               errs = append(errs, errors...)
+                       }
                }
-               errs = append(errs, errors...)
        }

        return adapterRequests, errs
```